symbolic-unwind/src/base.rs

```diff
@@ -0,0 +1,193 @@
+//! Basic definitions necessary for stack unwinding.
+use std::convert::TryInto;
+use std::fmt::Debug;
+use std::ops::{Add, Div, Mul, Rem, Sub};
+
+/// Trait that abstracts over the [endianness](https://en.wikipedia.org/wiki/Endianness)
+/// of data representation.
+///
+/// This trait provides no other functionality than a method for testing whether
+/// an endianness is big or little. In particular it does not provide methods for
+/// reading number types the way that similar traits/types in `byteorder` and `gimli` do.
+///
+/// We use our own trait here so we don't need to depend on crates (`byteorder`/`gimli`)
+/// for a minuscule part of their features.
+pub trait Endianness: Debug + Clone + Copy {
+    /// Returns true if this is big-endian (i.e. most significant bytes first).
+    fn is_big_endian(self) -> bool;
+}
+
+/// Big-endian data representation (i.e. most significant bits first),
+/// known at compile time.
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct BigEndian;
+
+impl Endianness for BigEndian {
+    fn is_big_endian(self) -> bool {
+        true
+    }
+}
+
+/// Little-endian data representation (i.e. least significant bits first),
+/// known at compile time.
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct LittleEndian;
+
+impl Endianness for LittleEndian {
+    fn is_big_endian(self) -> bool {
+        false
+    }
+}
+
+/// Endianness that can be selected at run time.
+///
+/// Defaults to the endianness of the target platform.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum RuntimeEndian {
+    /// Big-endian data representation.
+    Big,
+    /// Little-endian data representation.
+    Little,
+}
+
+impl Default for RuntimeEndian {
+    #[cfg(target_endian = "little")]
+    fn default() -> Self {
+        Self::Little
+    }
+
+    #[cfg(target_endian = "big")]
+    fn default() -> Self {
+        Self::Big
+    }
+}
+
+impl Endianness for RuntimeEndian {
+    fn is_big_endian(self) -> bool {
+        self == Self::Big
+    }
+}
+
+/// A trait for types that can be used as memory addresses.
+///
+/// This contains no actual functionality, it only bundles other traits.
+pub trait RegisterValue:
+    TryInto<usize>
+    + Add<Output = Self>
+    + Mul<Output = Self>
+    + Div<Output = Self>
+    + Sub<Output = Self>
+    + Rem<Output = Self>
+    + Copy
+    + Sized
+    + Debug
+{
+    /// The number of bytes that need to be read to produce one value of this type.
+    const WIDTH: usize;
+
+    /// Attempt to read a value of this type from a slice of bytes.
+    ///
+    /// May fail if an invalid byte is encountered or there are not enough bytes in the slice.
```

Comment:
Is there even such a thing as an invalid byte when we are dealing with binary data?